### PR TITLE
ci: Run clippy lint for undocumented unsafe code

### PR DESCRIPTION
**Pull request by pierwill**
The [`clippy::undocumented_unsafe_blocks`](https://rust-lang.github.io/rust-clippy/master/#undocumented_unsafe_blocks) lint is allowed by default. As a "restriction lint" it must be explicitly specified.

Turning this on will cause the following code to fail in CI

```rust
unsafe {
    [...snip...]
}
```

while allowing this:

```rust
// SAFETY: The following code is safe because [good reasons].
unsafe {
    [...snip...]
}
```

### CODE REVIEW

**Comment by pierwill** [comment]
[This](https://app.circleci.com/pipelines/github/influxdata/influxdb_iox/14477/workflows/a396217b-1057-4476-a53f-fd6bb67e2b1f/jobs/110010?invite=true#step-104-737) may actually be a bug in clippy...

**Comment by crepererum** [comment]
I think you might need to indent the line after the `SAFETY` word, but not sure. Also can you please instead of adding the lint via CLI add it to the individual crates, like:

https://github.com/influxdata/influxdb_iox/blob/7261571abfe1b6088810732d5811240ec585c7cf/ingester/src/lib.rs#L6-L13

**Comment by pierwill** [comment]
Asking about this now on the Zulip channel for clippy.

**Comment by pierwill** [comment]
This does *not* reproduce [on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f1f31ccdb2091ca583c025b5a6dfb655).

**Comment by pierwill** [comment]
Ah okay, I think the problem is with the version of clippy.

File: data_types/src/lib.rs [hunk 1]
```diff
@@ -2,7 +2,12 @@
 //! servers including replicated data, rules for how data is split up and
 //! queried, and what gets stored in the write buffer database.
 
-#![deny(rustdoc::broken_intra_doc_links, rustdoc::bare_urls, rust_2018_idioms)]
```
Comment (crepererum) [inline data_types/src/lib.rs:5]: Could you please add this to ALL crates? I know it's a bit cumbersome but sadly there isn't any workspace-wide way to declare `allow`/`warn`/`deny` lints.

Comment (pierwill) [reply data_types/src/lib.rs:5]: Sure thing 👍 . 

However, I'm still investigating the the mysterious failure of this lint to work on this repo only....

**Comment by pierwill** [comment]
This is an incredibly frustrating bug. 😠

**Comment by alamb** [comment]
marking as a draft so it is clear it isn't actively waiting for a review

**Comment by alamb** [comment]
This PR seems to have bitrotted -- closing the PR to clean up our review queue. Feel free to reopen when ready for review again

**Pull request closed without merging**